start review of first part (#2)
zkamvar/user2021: prose.md

> timing 2021-06-10: 04:01.98

The Carpentries is a global organization of volunteers who teach
foundational data and coding skills to researchers worldwide. What makes us
stand out is not the fact that we use evidence-based practices in teaching, it's
that we align all of our decisions with our core values: we put people first
and we value all contributions.

Our lessons are distributed as open-licensed websites built on a consistent style that
adheres to active learning principals. We use these lessons as the source

> material for the hundreds of workshops we run each year. So, these sites have three distinct

Reviewer: I was thinking about how awkward my transition here was and I think your solution for it works perfectly in a speaking context :clap:

audiences:

> 1. The Carpentries Instructors who refer to these materials as

Reviewer: I think it's good to highlight that we do certify the instructors to give a little more weight to their role. 

```suggestion
1. The certified Carpentries Instructors who refer to these materials as
```

   they teach our workshops.
2. The learners in a Carpentries workshop who rely on these lessons after a workshop
   as they review and practice their newly aquired skills.
3. The educators who adapt these open-licensed materials for their own
   lessons.

> The source for these lessons are hosted on GitHub were

Reviewer: ```suggestion
The source for these lessons are hosted on GitHub where
```

Reviewer: That was fully my fault

further our mission**. I will introduce you to our current
infrastructure, its unique challenges, our solutions, and how we used past and
present feedback from our community to iteratively refine our design.

Before continuing, I want to remind everyone of two things. First:

> There is no right or wrong, only better or worse.
>
> --- GW

Greg Wilson wrote this after several iterations of the lesson
infrastructure. I am putting it here as a reminder that the infrastructure we have
now was working for us at the time. What we come up with will have its
own difficulties down the road, but what is important is that we build something
that better addresses the needs of our community, which leads me to my second
reminder:

> In The Carpentries: you belong

As the community has grown, our infrastructure has been put to the test and we
have continuously updated our workflows to make it easier for people to contribute.
The reason we do this is because our values are what drive us. In The Carpentries, you
belong, no matter if you have been working as a systems administrator for a
university HPC cluster or if you have just learned to write your first R script.

> To better understand the decisions we are making, let's start by reviewing

Reviewer: :100:

our current infrastructure.

## Infrastructure

> timing 2021-06-09: 04:20.38
> timing 2021-06-09: 03:52.10
> timing 2021-06-10: 03:08.87

Our lessons are written in markdown and into a website via GitHub Pages and
the Jekyll static site generator. The idea behind this choice was that it was
the most straightforward way to create a static website without needing a server
like Wordpress or Drupal sites. Ideally, it also provided a way for people to
use this as an example of how they could build their own website.

The paradigm of being able to write markdown lessons and get a functional website
out of it is not a new concept. In fact, there are over 460 iterations of
this concept according to https://staticsitegenerators.net. Jekyll happens to
be the one that GitHub implemented early on to provide documentation for the

> open-source project it hosted, and thus, it stuck.

Reviewer: ah, good point about highlighting the _reason_ why they chose it. Is the project you are referring to github itself?

1. Installation pains.
   <https://twitter.com/gvwilson/status/1380115181708189701?s=20> Having
   Jekyll, Python, and Make as dependencies means that people who want to build
   these lessons need all three of these successfully installed and up-to-date on
   their machines. This is especially frustrating for Windows users who have none
   of these by default. Moreover, having all the tools living inside the
   repository meant that lessons quickly become out of date as we improve our
   infrastructure.
   these lessons on their machines need all three of these successfully installed 
   and up-to-date.
   This is especially frustrating for Windows users who have none
   of these by default.
2. All the scripts live inside the repository. This means that lessons quickly
   become out of date as they diverge from the improvements made upstream.
2. Another drawback is that we have a lesson website wrapped around a static
   site generator (which in and of itself is a kind of desire path), this meant
   that it was easy to contribute to if you were familiar with how Jekyll
   operated. Some would say Jekyll is well documented, but in my experience,

>    Jekyll is well, documented. There is often a moment of panic in a new

Reviewer: Aw, this one of my favorite parts!

>    trail head that ended up built in different ways.

Reviewer: ```suggestion
   paths that ended up built in different ways.
```

> begun finding the trail head of these desire paths all across our lesson
> infrastructure.

Reviewer: ```suggestion
begun finding these desire paths all across our lesson infrastructure.
```

> This manifested through our communication channels and frustration from contributors

Reviewer: ```suggestion
These manifested through our communication channels and frustration from contributors
```

> and maintainers alike. The thing about desire paths,

Reviewer: I think we should avoid mixing these metaphors since trail heads are associated with official paths.

> infrastructure?

Reviewer: Note: add mention of links to docs, blog posts, etc.

> How and why did we choose R to create the next iteration of our lesson
> infrastructure?

Reviewer: ```suggestion
The solution we came up with uses R and I simply do not have enough time to discuss all the features, but you can find out for yourself if you visit our blog posts and documentation on how to get started. 

So, how and why did we choose R to create the next iteration of our lesson infrastructure?
```

out several static site generators, but the largest barrier for many of these
was that they were not easy to install and maintain.

We settled on the fact that R is the best place to go for our needs
because:

1. First and foremost: R is chock-full of friendly communities!
  - RLadies
  - ROpenSci
  - RForwards
  - MiR
2. R is easy to install on all major operating systems and
   is available online via RStudio Cloud
3. R has a robust ecosystem for publishing thanks to {knitr} and RMarkdown

Once we identified R as our solution, the natural place to go was one of the
RMarkdown variants like {blogdown} or {hugodown}, but we found that while the tools
were indeed separated from the content and the documentation was rich and
accessible, there were many aspects that would not fit our needs including the
presence of styling within the repository.

We realized that an unlikely contender, {pkgdown}, a documentation site
generator, used the same model that we were looking for: It meets people where
they are, content is pure markdown with no extra templating required, the tools
to build everything lived in a separate package, and you can customize its appearance

> by making your own package to supply your CSS, JS, and HTML templates.

Reviewer: every time i see supply, I think it's a misspelling of `sapply()` and I want to get rid of it. 

R has truly rotted my brain.

 - {varnish} to host the CSS, JS, and HTML templates
 - {pegboard} to serve as a validator (and converter!) for lesson content

When thinking about how that compose the lessons should be organized,

> We created a folder structure that reflects the website

Reviewer: I think this transition is a bit cumbersome in both versions. :thinking: what is the best way to convey that we are focusing on the way people view the lesson itself?

> When thinking about how that compose the lessons should be organized,
> We created a folder structure that reflects the website

Reviewer: ```suggestion
Of course, the first thing any contributor will see is the folder structure. We organized the folders of the lesson to reflect the layout of website
```

The final folder is one of the access panels that contains the rendered markdown
files---so you can use them in another context---and static website so that you
can put it on a USB stick and share it without needing to run a local server.

### Opportunity

Our solution ticks all the boxes that satisfy our design choices,
but if we ended up designing something that R enthusiasts
love, but is unusable by newcomers, Python, or Matlab folks, then we are not
in line with our value of being inclusive of all.

Our goal is for authors to focus on the content over the tooling.
We want lesson authors to create a lesson directly from the source
they already work with: markdown/RMarkdown/MyST notebooks.
To do this, we needed to test the minimal
viable product on actual maintainers and we needed to make sure that they were
spread across the spectrum of familiarity with:

 - using R,
 - the current infrastructure,
 - the way The Carpentries operates.

We recruited a total of 19 volunteers to run through Alpha Testing, which
tested the participants ability to install the required software and packages,
create,
modify, and contribute to lessons. After the tests, I

> interviewed for 20 minutes volunteers about their experience

Reviewer: ```suggestion
asked for 20-minute open-ended interviews from the volunteers about their experience
```

> to identify the features people kept stumbling over. I want to take a

Reviewer: ```suggestion
to identify common stumbling blocks, challenges, and bright points. I want to take a
```

> timing 2021-06-09: ~ 1 minute
> timing 2021-06-10: ~ 1 minute

We have just finished with this first testing phase and the next step for us is
to address the questions and concerns that were brought up (for example: how
do I use this without clobbering the current R installation on my system?),
improve the documentation, and get ready for the beta release, where we will try it
on a few live lessons to identify painpoints for the community. It is a slow
process, but this way we can avoid major unforseen issues (minor issues are a
given), bring users in on the ground floor, get valuable feedback, and
strengthen trust with our community.

> > if you want to go fast, go alone; if you want to go far, go together

Reviewer: I was low-key looking to avoid this quote partially because it's cheesy, but also because there is an aspect of this that's only partially true of this project. Yes, we are gathering community feedback for it, but the vast majority of the project was written by myself.

And I want to conclude by saying that we ended up choosing a solution that we

> believe aligns with our values and will not get in the way of the work from our community.

Reviewer: ```suggestion
believe aligns with our values and will work with our community.
```
